src/goto-instrument/synthesizer/synthesizer_utils.cpp

```diff
@@ -112,3 +117,96 @@ void annotate_invariants(
       invariant_map_entry.second;
   }
 }
+
+bool is_instruction_in_transfomed_loop(
+  const loop_idt &loop_id,
+  const goto_functiont &function,
+  unsigned location_number_of_target)
+{
+  // The loop body after transformation are instructions from
+  // loop havocing instructions
+  // to
+  // loop end of transformed code.
+
+  bool result = false;
+  unsigned location_number_of_havocing;
+  for(goto_programt::const_targett it = function.body.instructions.begin();
+      it != function.body.instructions.end();
+      ++it)
+  {
+    if(is_loop_havoc(it) && it->loop_number == loop_id.loop_number)
+    {
+      location_number_of_havocing = it->location_number;
+    }
+
+    if(is_transformed_loop_end(it) && it->loop_number == loop_id.loop_number)
+    {
+      result =
+        result || (location_number_of_havocing < location_number_of_target &&
+                   location_number_of_target < it->location_number);
+    }
+  }
+
+  return result;
+}
+
+std::size_t extract_size(const exprt &expr)
+{
+  const exprt *node = &expr;
+
+  while(!can_cast_expr<symbol_exprt>(*node))
+  {
+    node = &node->operands()[0];
+  }
+
+  const auto &symbol_expr = expr_try_dynamic_cast<symbol_exprt>(*node);
+  INVARIANT(
+    can_cast_type<array_typet>(symbol_expr->type()),
+    "Dynamic object has array type.");
+
+  const auto &type = type_try_dynamic_cast<array_typet>(symbol_expr->type());
+
+  return hex_to_size_t(type->size().get_string(ID_value));
+}
+
+std::size_t extract_offset(const exprt &expr)
+{
+  const exprt *node = &expr;
+
+  while(!can_cast_expr<plus_exprt>(*node))
+  {
+    if(node->operands().size() == 0)
+      return 0;
+    node = &node->operands()[0];
+  }
+
+  return hex_to_size_t(node->operands()[1].get_string(ID_value));
+}
+
+std::size_t hex_to_size_t(const std::string &hex_str)
+{
+  std::istringstream converter(hex_str);
+  size_t result;
+  converter >> std::hex >> result;
+  return result;
+}
+
+invariant_mapt combine_in_and_post_invariant_clauses(
+  invariant_mapt in_clauses,
+  invariant_mapt post_clauses,
+  invariant_mapt neg_guards)
```

Comment:
Should all of these be const references?

Reply:
Yes, fixed.